### DISCUSSION TITLE
Deploy--->alex--->dev--->managed-learn--->ml-reports-service

### DISCUSSION THREAD

**Discussion by srikanth-6520** [Sunbird-Ed/Community#532] [Bugs]
Hi Team,
we are getting an error, while deploying the ml-report-service jenkins job, below we have provided jenkins console.

Started by user [Kolluri Srikanth]
Rebuilds build #[2](https://jenkins.alex-sb.dev.verlab.io/job/Deploy/job/alex/job/dev/job/managed-learn/job/ml-reports-service/2)
Checking out git https://github.com/project-sunbird/sunbird-devops.git into /var/lib/jenkins/workspace/Deploy/alex/dev/managed-learn/ml-reports-service@script/74766eba9651ee97692d587da848d14cae2c225f81df644eb5a529f729d5e85a to read kubernetes/pipelines/deploy_core/Jenkinsfile
The recommended git tool is: NONE
No credentials specified
 > git rev-parse --resolve-git-dir /var/lib/jenkins/workspace/Deploy/alex/dev/managed-learn/ml-reports-service@script/74766eba9651ee97692d587da848d14cae2c225f81df644eb5a529f729d5e85a/.git # timeout=10
Fetching changes from the remote Git repository
 > git config remote.origin.url https://github.com/project-sunbird/sunbird-devops.git # timeout=10
Using shallow fetch with depth 1
Fetching upstream changes from https://github.com/project-sunbird/sunbird-devops.git
 > git --version # timeout=10
 > git --version # 'git version 2.17.1'
 > git fetch --tags --progress --depth=1 -- https://github.com/project-sunbird/sunbird-devops.git +refs/heads/*:refs/remotes/origin/* # timeout=10
 > git rev-parse origin/release-5.1.0^{commit} # timeout=10
Checking out Revision bee2ee6f39baa087eba22646d7919493552b10e1 (origin/release-5.1.0)
 > git config core.sparsecheckout # timeout=10
 > git checkout -f bee2ee6f39baa087eba22646d7919493552b10e1 # timeout=10
Commit message: "Merge pull request #3793 from project-sunbird/gandham-santhosh-patch-1"
First time build. Skipping changelog.
Loading library deploy-conf@heads/shared-lib
Attempting to resolve heads/shared-lib from remote references...
 > git --version # timeout=10
 > git --version # 'git version 2.17.1'
 > git ls-remote -- https://github.com/project-sunbird/sunbird-devops.git # timeout=10
The recommended git tool is: NONE
No credentials specified
 > git rev-parse --resolve-git-dir /var/lib/jenkins/workspace/Deploy/alex/dev/managed-learn/ml-reports-service@libs/b4ac4d1235c67cfedda9d7bf40936b35bc6929460ef19107abe1b183748d4145/.git # timeout=10
Fetching changes from the remote Git repository
 > git config remote.origin.url https://github.com/project-sunbird/sunbird-devops.git # timeout=10
Using shallow fetch with depth 1
Fetching upstream changes from https://github.com/project-sunbird/sunbird-devops.git
 > git --version # timeout=10
 > git --version # 'git version 2.17.1'
 > git fetch --no-tags --progress --depth=1 -- https://github.com/project-sunbird/sunbird-devops.git +refs/heads/*:refs/remotes/origin/* refs/heads/shared-lib # timeout=10
Checking out Revision 778da7ce7f97efd003eed98071921a837d209343 (778da7ce7f97efd003eed98071921a837d209343)
 > git config core.sparsecheckout # timeout=10
 > git checkout -f 778da7ce7f97efd003eed98071921a837d209343 # timeout=10
Commit message: "fix: move variables to be accessible across both try blocks"
 > git rev-list --no-walk 778da7ce7f97efd003eed98071921a837d209343 # timeout=10
[Pipeline] Start of Pipeline
[Pipeline] node
Running on [Jenkins](https://jenkins.alex-sb.dev.verlab.io/computer/(built-in)/) in /var/lib/jenkins/workspace/Deploy/alex/dev/managed-learn/ml-reports-service
[Pipeline] {
[Pipeline] stage
[Pipeline] { (checkout public repo)
[Pipeline] echo
Found .git folder. Clearing it..
[Pipeline] sh
+ git clean -fxd
Removing ansible/inventory/env/hosts
Removing ansible/inventory/env/keys/
Removing ansible/inventory/env/secrets.yml
Removing kubernetes/helm_charts/core/ml-reports-service/values.yaml
Removing metadata.json
Skipping repository private/
Removing private@tmp/
[Pipeline] checkout
The recommended git tool is: NONE
No credentials specified
 > git rev-parse --resolve-git-dir /var/lib/jenkins/workspace/Deploy/alex/dev/managed-learn/ml-reports-service/.git # timeout=10
Fetching changes from the remote Git repository
 > git config remote.origin.url https://github.com/project-sunbird/sunbird-devops.git # timeout=10
Using shallow fetch with depth 1
Fetching upstream changes from https://github.com/project-sunbird/sunbird-devops.git
 > git --version # timeout=10
 > git --version # 'git version 2.17.1'
 > git fetch --tags --progress --depth=1 -- https://github.com/project-sunbird/sunbird-devops.git +refs/heads/*:refs/remotes/origin/* # timeout=10
 > git rev-parse origin/release-5.1.0^{commit} # timeout=10
Checking out Revision bee2ee6f39baa087eba22646d7919493552b10e1 (origin/release-5.1.0)
 > git config core.sparsecheckout # timeout=10
 > git checkout -f bee2ee6f39baa087eba22646d7919493552b10e1 # timeout=10
Commit message: "Merge pull request #3793 from project-sunbird/gandham-santhosh-patch-1"
[Pipeline] }
[Pipeline] // stage
[Pipeline] stage
[Pipeline] { (deploy)
[Pipeline] ansiColor
[Pipeline] {

[Pipeline] sh
+ echo Deploy/alex/dev/managed-learn/ml-reports-service
[Pipeline] sh
+ echo Deploy/alex/dev/managed-learn/ml-reports-service
[Pipeline] sh
+ echo Deploy/alex/dev/managed-learn/ml-reports-service
[Pipeline] stage
[Pipeline] { (parameter checks)
[Pipeline] echo
Found environment variable named hub_org with value as: 961571523562.dkr.ecr.ap-south-1.amazonaws.com/prod
[Pipeline] copyArtifacts
Copied 1 artifact from "[ArtifactUpload » dev » managed-learn » ml-core-service](https://jenkins.alex-sb.dev.verlab.io/job/ArtifactUpload/job/dev/job/managed-learn/job/ml-core-service/)" build number [3](https://jenkins.alex-sb.dev.verlab.io/job/ArtifactUpload/job/dev/job/managed-learn/job/ml-core-service/3/)
[Pipeline] sh
+ jq -r .image_name metadata.json
[Pipeline] sh
+ jq -r .node_name metadata.json
[Pipeline] echo
image_tag not specified, using the image_tag specified in metadata.json. 
[Pipeline] sh
+ jq -r .image_tag metadata.json
[Pipeline] }
[Pipeline] // stage
[Pipeline] }

[Pipeline] // ansiColor
[Pipeline] sh
+ echo Deploy/alex/dev/managed-learn/ml-reports-service
[Pipeline] sh
+ pwd
[Pipeline] ansiColor
[Pipeline] {

[Pipeline] stage
[Pipeline] { (ansible-run)
[Pipeline] echo
{env=dev, module=managed-learn, jobName=ml-reports-service, absolute_job_path=ArtifactUpload/dev/managed-learn/ml-core-service, agent=built-in, image_name=ml-core-service, image_tag=release-5.1.0_RC6_3cde99f_1, currentWs=/var/lib/jenkins/workspace/Deploy/alex/dev/managed-learn/ml-reports-service, ansiblePlaybook=/var/lib/jenkins/workspace/Deploy/alex/dev/managed-learn/ml-reports-service/kubernetes/ansible/deploy_core_service.yml, ansibleExtraArgs=--extra-vars "chart_path=/var/lib/jenkins/workspace/Deploy/alex/dev/managed-learn/ml-reports-service/kubernetes/helm_charts/core/ml-reports-service release_name=ml-reports-service image_tag=release-5.1.0_RC6_3cde99f_1 image_name=ml-core-service role_name=sunbird-deploy" --vault-password-file /var/lib/jenkins/secrets/vault-pass}
[Pipeline] dir
Running in /var/lib/jenkins/workspace/Deploy/alex/dev/managed-learn/ml-reports-service/private
[Pipeline] {
[Pipeline] echo
Info: Branch override is enabled
[Pipeline] checkout
The recommended git tool is: NONE
using credential bitbucket-cred
 > git rev-parse --resolve-git-dir /var/lib/jenkins/workspace/Deploy/alex/dev/managed-learn/ml-reports-service/private/.git # timeout=10
Fetching changes from the remote Git repository
 > git config remote.origin.url bitbucket.org:verlab/veranda-alex-sb-private.git # timeout=10
Using shallow fetch with depth 1
Fetching upstream changes from bitbucket.org:verlab/veranda-alex-sb-private.git
 > git --version # timeout=10
 > git --version # 'git version 2.17.1'
using GIT_SSH to set credentials Veranda Bitbucket Credential (https://bitbucket.org/verlab/veranda-alex-sb-private)
Verifying host key using known hosts file
 > git fetch --no-tags --progress --depth=1 -- bitbucket.org:verlab/veranda-alex-sb-private.git +refs/heads/*:refs/remotes/origin/* # timeout=10
 > git rev-parse origin/release-5.1.0^{commit} # timeout=10
Checking out Revision c21be57697b421f3779459deb0cd8ae852820815 (origin/release-5.1.0)
 > git config core.sparsecheckout # timeout=10
 > git checkout -f c21be57697b421f3779459deb0cd8ae852820815 # timeout=10
Commit message: "we have added "druid" var"
 > git rev-list --no-walk 181df246f5e15f714a812b34f890d4b7b51a4bb9 # timeout=10
[Pipeline] }
[Pipeline] // dir
[Pipeline] sh
+ rsync -Lkr /var/lib/jenkins/workspace/Deploy/alex/dev/managed-learn/ml-reports-service/private/ansible/inventory/dev/managed-learn/common.yml /var/lib/jenkins/workspace/Deploy/alex/dev/managed-learn/ml-reports-service/private/ansible/inventory/dev/managed-learn/hosts /var/lib/jenkins/workspace/Deploy/alex/dev/managed-learn/ml-reports-service/private/ansible/inventory/dev/managed-learn/keys /var/lib/jenkins/workspace/Deploy/alex/dev/managed-learn/ml-reports-service/private/ansible/inventory/dev/managed-learn/secrets.yml /var/lib/jenkins/workspace/Deploy/alex/dev/managed-learn/ml-reports-service/ansible/inventory/env/
+ [ -f /var/lib/jenkins/workspace/Deploy/alex/dev/managed-learn/ml-reports-service/ansible/inventory/env/kubernetes.yaml ]
+ ansible-playbook -i /var/lib/jenkins/workspace/Deploy/alex/dev/managed-learn/ml-reports-service/ansible/inventory/env/hosts /var/lib/jenkins/workspace/Deploy/alex/dev/managed-learn/ml-reports-service/kubernetes/ansible/deploy_core_service.yml --extra-vars chart_path=/var/lib/jenkins/workspace/Deploy/alex/dev/managed-learn/ml-reports-service/kubernetes/helm_charts/core/ml-reports-service release_name=ml-reports-service image_tag=release-5.1.0_RC6_3cde99f_1 image_name=ml-core-service role_name=sunbird-deploy --vault-password-file /var/lib/jenkins/secrets/vault-pass
[WARNING]: Invalid characters were found in group names but not replaced, use
-vvvv to see details

[WARNING]:  * Failed to parse
/var/lib/jenkins/workspace/Deploy/alex/dev/managed-learn/ml-reports-
service/ansible/inventory/env/hosts with yaml plugin: Syntax Error while
loading YAML.   found unexpected ':'  The error appears to be in
'/var/lib/jenkins/workspace/Deploy/alex/dev/managed-learn/ml-reports-
service/ansible/inventory/env/hosts': line 1, column 5, but may be elsewhere in
the file depending on the exact syntax problem.  The offending line appears to
be:   [all:vars]     ^ here

[WARNING]:  * Failed to parse
/var/lib/jenkins/workspace/Deploy/alex/dev/managed-learn/ml-reports-
service/ansible/inventory/env/hosts with ini plugin:
/var/lib/jenkins/workspace/Deploy/alex/dev/managed-learn/ml-reports-
service/ansible/inventory/env/hosts:150: Section [core:children] includes
undefined group: postgresql-slave

[WARNING]: Unable to parse /var/lib/jenkins/workspace/Deploy/alex/dev/managed-
learn/ml-reports-service/ansible/inventory/env/hosts as an inventory source

[WARNING]: No inventory was parsed, only implicit localhost is available

[WARNING]: provided hosts list is empty, only localhost is available. Note that
the implicit localhost does not match 'all'

[WARNING]: While constructing a mapping from
/var/lib/jenkins/workspace/Deploy/alex/dev/managed-learn/ml-reports-
service/kubernetes/ansible/roles/sunbird-deploy/defaults/main.yml, line 3,
column 1, found a duplicate dict key (sunbird_time_zone). Using last defined
value only.

[WARNING]: While constructing a mapping from
/var/lib/jenkins/workspace/Deploy/alex/dev/managed-learn/ml-reports-
service/kubernetes/ansible/roles/sunbird-deploy/defaults/main.yml, line 3,
column 1, found a duplicate dict key (nginx_per_ip_connection_limit). Using
last defined value only.

[WARNING]: While constructing a mapping from
/var/lib/jenkins/workspace/Deploy/alex/dev/managed-learn/ml-reports-
service/kubernetes/ansible/roles/sunbird-deploy/defaults/main.yml, line 3,
column 1, found a duplicate dict key (sunbird_portal_player_cdn_enabled). Using
last defined value only.

[WARNING]: While constructing a mapping from
/var/lib/jenkins/workspace/Deploy/alex/dev/managed-learn/ml-reports-
service/kubernetes/ansible/roles/sunbird-deploy/defaults/main.yml, line 3,
column 1, found a duplicate dict key (dial_liveness_readiness). Using last
defined value only.

[WARNING]: While constructing a mapping from
/var/lib/jenkins/workspace/Deploy/alex/dev/managed-learn/ml-reports-
service/kubernetes/ansible/roles/sunbird-deploy/defaults/main.yml, line 924,
column 5, found a duplicate dict key (failureThreshold). Using last defined
value only.

[WARNING]: While constructing a mapping from
/var/lib/jenkins/workspace/Deploy/alex/dev/managed-learn/ml-reports-
service/kubernetes/ansible/roles/sunbird-deploy/defaults/main.yml, line 934,
column 5, found a duplicate dict key (failureThreshold). Using last defined
value only.

[WARNING]: While constructing a mapping from
/var/lib/jenkins/workspace/Deploy/alex/dev/managed-learn/ml-reports-
service/kubernetes/ansible/roles/sunbird-deploy/defaults/main.yml, line 947,
column 5, found a duplicate dict key (failureThreshold). Using last defined
value only.

[WARNING]: While constructing a mapping from
/var/lib/jenkins/workspace/Deploy/alex/dev/managed-learn/ml-reports-
service/kubernetes/ansible/roles/sunbird-deploy/defaults/main.yml, line 957,
column 5, found a duplicate dict key (failureThreshold). Using last defined
value only.

PLAY [local] *******************************************************************

TASK [check for kong db migration] *********************************************
skipping: [localhost]

TASK [sunbird-deploy : template values.yaml file] ******************************
[DEPRECATION WARNING]: Distribution Ubuntu 18.04 on host localhost should use 
/usr/bin/python3, but is using /usr/bin/python for backward compatibility with 
prior Ansible releases. A future Ansible release will default to using the 
discovered platform python for this host. See [https://docs.ansible.com/ansible/](https://docs.ansible.com/ansible/%1B[0m)
2.8/reference_appendices/interpreter_discovery.html for more information. This 
feature will be removed in version 2.12. Deprecation warnings can be disabled 
by setting deprecation_warnings=False in ansible.cfg.
changed: [localhost]

TASK [sunbird-deploy : copy and template env file from swarm role] *************
failed: [localhost] (item=../../../../ansible/roles/stack-sunbird/templates/ml-reports-service.env) => {
    "ansible_loop_var": "item",
    "changed": false,
    "item": "../../../../ansible/roles/stack-sunbird/templates/ml-reports-service.env"
}

MSG:

AnsibleUndefinedVariable: 'dict object' has no attribute 'druid'

PLAY RECAP *********************************************************************
localhost                  : ok=1    changed=1    unreachable=0    failed=1    skipped=1    rescued=0    ignored=0   

[Pipeline] }
[Pipeline] // stage
[Pipeline] }

[Pipeline] // ansiColor
[Pipeline] }
[Pipeline] // stage
[Pipeline] ansiColor
[Pipeline] {

[Pipeline] stage
[Pipeline] { (slack_notify)
[Pipeline] sh
+ git rev-parse --short HEAD
[Pipeline] sh
+ git name-rev --name-only HEAD
+ rev
+ cut -d / -f1
+ rev
[Pipeline] sh
+ echo Deploy/alex/dev/managed-learn/ml-reports-service
[Pipeline] echo
Could not find env specific Slack channel. Check for global slack channel..
[Pipeline] echo
Could not find slack environment variable. Skipping slack notification..
[Pipeline] }
[Pipeline] // stage
[Pipeline] }

[Pipeline] // ansiColor
[Pipeline] ansiColor
[Pipeline] {

[Pipeline] stage
[Pipeline] { (email_notify)
[Pipeline] sh
+ echo Deploy/alex/dev/managed-learn/ml-reports-service
[Pipeline] echo
Could not find env specific email group. Check for global email group..
[Pipeline] echo
Could not find global email group variable. Skipping email notification..
[Pipeline] }
[Pipeline] // stage
[Pipeline] }

[Pipeline] // ansiColor
[Pipeline] }
[Pipeline] // node
[Pipeline] End of Pipeline
ERROR: script returned exit code 2
Finished: FAILURE

**Comment by tsprasath** [comment]
@srikanth-6520 Please check the variable ```druid``` present in the private repository.

**Reply by srikanth-6520** [reply]
@tsprasath druid variable present in the private repo with an empty value.

**Reply by tsprasath** [reply]
[aks30](https://github.com/aks30) Kindly help @srikanth-6520 on ML service

**Reply by heungheung** [reply]
in your private repo, inventory/dev/Core/hosts, point the druid host to the corresponding IP, e.g.
```
[druid]
10.0.1.9
```

managed-lern uses soft link to Core, that's why you will need to have values in Core. The error message already provide the cause,

https://github.com/project-sunbird/sunbird-devops/blob/release-5.1.0/ansible/roles/stack-sunbird/templates/ml-reports-service.env#L9

will need to get the IP address of druid and use it during the deployment for setting up the druid URL, i.e. default(groups['druid'][0]) - the first IP address under the group druid. 

For development setup (6 VM), all druid components will be in a single VM and all you need is to put it in the hosts file of Core

**Reply by srikanth-6520** [reply]
@aks30 we have updated druid ip in core hosts , and redeployed the job , jenkins job got success , but the service is crashlooping in k8s, we have check the service logs , there we found the service error is "Server could not start , Not all the environment is provided" , and we have check the configmap of the service , there we can see all the variables are provide , for you are reference we have provided the configmap.

apiVersion: v1
data:
  APPLICATION_ENV: alexdev
  APPLICATION_PORT: "3000"
  CONTENT_REPORT_THRESHOLD: "5"
  DRUID_URL: http://172.16.1.193:8082
  ENTITY_SCORE_REPORT_THRESHOLD: "5"
  EVIDENCE_THRESHOLD: "3"
  GOTENBERG_URL: http://gotenberg-service:3000
  INTERNAL_ACCESS_TOKEN: 85f6cc767b7db818e2f9
  KEYCLOAK_PUBLIC_KEY_PATH: keycloak-public-keys
  ML_CORE_SERVICE_URL: http://ml-core-service:3000
  ML_SURVEY_SERVICE_URL: http://ml-survey-service:3000
  OBSERVATION_DATASOURCE_NAME: sl-observation
  OBSERVATION_EVIDENCE_DATASOURCE_NAME: sl-observation-evidence
  OBSERVATION_SCORE_REPORT_THRESHOLD: "2"
  SURVEY_DATASOURCE_NAME: sl-survey
  SURVEY_EVIDENCE_DATASOURCE_NAME: sl-survey-evidence

**Reply by aks30** [reply]
@srikanth-6520 Please suggest which tag is being used to deploy and any logs we could capture in the kubernetes pod

**Reply by srikanth-6520** [reply]
@aks30 we have used "release-5.1.0" tag, and there is no logs we are getting in k8s pod , apart from provided the above logs .

**Reply by ankit1shahu** [reply]
@srikanth-6520, could you please provide the logs from Rancher or Graylog of ml-reports? Additionally, can you confirm whether the variables in your .env file are defined using the format APPLICATION_ENV: alexdev or APPLICATION_ENV= alexdev?

**Reply by Krishnaj20** [reply]
@srikanth-6520 - Is your query addressed?

**Reply by srikanth-6520** [reply]
yes @Krishnaj20 , it was fixed.
